synapse/http/client.py

```diff
@@ -83,18 +85,27 @@ def __init__(self, hs):
         if hs.config.user_agent_suffix:
             self.user_agent = "%s %s" % (self.user_agent, hs.config.user_agent_suffix,)
 
+        self.user_agent = self.user_agent.encode('ascii')
+
     @defer.inlineCallbacks
     def request(self, method, uri, *args, **kwargs):
         # A small wrapper around self.agent.request() so we can easily attach
         # counters to it
         outgoing_requests_counter.labels(method).inc()
 
         # log request but strip `access_token` (AS requests for example include this)
-        logger.info("Sending request %s %s", method, redact_uri(uri))
+        logger.info("Sending request %s %s", method, redact_uri(uri.encode('ascii')))
 
         try:
-            request_deferred = self.agent.request(
-                method, uri, *args, **kwargs
+            if "data" in kwargs:
+                data = kwargs.pop("data")
+            elif "bodyProducer" in kwargs:
+                data = kwargs.pop("bodyProducer")
+            else:
+                data = None
+
+            request_deferred = treq.request(
+                method, uri, *args, agent=self.agent, data=data, **kwargs
```

Comment:
I think at this point its going to be cleaner if we get rid of `*args` and `**kwargs` and be explicit in what we expect, which I think is just `headers` and `data`.